### PR TITLE
API Summary Length

### LINKED ISSUE #9244: API must enforce maximum length for summaries for Guilds and Challenges
Guilds and Challenges have a summary field and the website enforces a maximum length for it:
https://github.com/HabitRPG/habitica/blob/6838b7d0a60c1b5b089ae5dab17f9cf10e09b937/website/common/script/constants.js#L9-L10

The API's routes for creating and editing Guilds and Challenges should enforce the same maximum.

Note that currently the API does not make the summary field mandatory - that is desired behaviour at the moment because a mandatory summary would be a breaking change for the API. Because of this, the API should enforce the maximum length only if a summary is provided. Don't make the summary mandatory.

There should also be tests that ensure that IF a summary is provided, it is not too long.

### PR DESCRIPTION
**Pull request by thewilloftheshadow**
[//]: # (Note: See http://habitica.fandom.com/wiki/Using_Your_Local_Install_to_Modify_Habitica%27s_Website_and_API for more info)

[//]: # (Put Issue # here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #9244

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)
API now rejects both new challenge and challenges update if summary is longer than maximum length
Adds tests for this new check

[//]: # (Put User ID in here - found on the Habitica website at User Icon > Settings > API)

----
UUID: 2eb203f5-1b1e-4e5d-84a4-22f130a1090e

### CODE REVIEW

**Comment by thewilloftheshadow** [comment]
It seems the linter got a bunch of stuff that wasn’t related to this... I’ll revert those

**Comment by thewilloftheshadow** [comment]
Can't get tests to run on my machine so I'm not really sure why these are failing right now

File: website/server/controllers/api-v3/challenges.js [hunk 1]
```diff
@@ -200,6 +205,7 @@ api.createChallenge = {
     const { user } = res.locals;
 
     req.checkBody('group', apiError('groupIdRequired')).notEmpty();
+    if (req.body.challenge.summary && req.body.challenge.summary.length > MAX_SUMMARY_SIZE_FOR_CHALLENGES) throw new BadRequest(res.t('summaryTooLong'));
```
Comment (paglias) [inline website/server/controllers/api-v3/challenges.js:208]: what if `req.body.challenge` is not defined? Also this cose should go after the validation errors

**Comment by paglias** [comment]
I would move the check to the model with a validator https://mongoosejs.com/docs/validation.html

**Comment by paglias** [comment]
@thewilloftheshadow the tests are failing because you're not checking for the existence of `req.body.challenge`, 

```
  fullError: TypeError: Cannot read property 'summary' of undefined
      at fn (/website/server/controllers/api-v3/challenges.js:208:28)
```

**Comment by thewilloftheshadow** [comment]
Got it 👍🏻
This was just my initial test to see if it would catch that field, but I couldn't ever see the full errors from my machine

**Comment by paglias** [comment]
To see the error you have to:

- In one terminal run `NODE_ENV=test npm start`
- In another one do `npm run test:api-v3:integration:separate-server` and also set `.only` on the failing test so only that one will run, then you'll see the error logs in the first terminal window

**Comment by thewilloftheshadow** [comment]
Ahh that's exactly what I needed! Tyty

**Comment by SabreCat** [comment]
Closing this PR as abandoned. Worth peeking at for whoever picks this up next, though!

**Pull request closed without merging**